pkg/migrations/sql/idempotent/exemplar.sql

```diff
@@ -0,0 +1,122 @@
+-- get_exemplar_label_positions returns the position of label_keys as a one-to-one mapping with label_keys. It returns
+-- the positions of all label keys corresponding to that metric, so that it remains easier to add null values to those indexes
+-- whose labels are not present in the exemplar being inserted at the golang level.
+CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.get_or_create_exemplar_label_key_positions(metric_name_text TEXT, label_keys TEXT[])
+RETURNS TABLE (metric_family_text TEXT, label_positions_map JSON) AS
+$$
+DECLARE
+    current_position INTEGER := 1; -- index in postgres starts from 1. Let's maintain the convention for less confusion.
+    new_position INTEGER;
+    k TEXT;
+BEGIN
+    LOCK TABLE SCHEMA_CATALOG.exemplar_label_key_position IN ACCESS EXCLUSIVE MODE;
+    -- If there isn't any data for the given metric_name_text.
+    IF ( SELECT count(key) = 0 FROM SCHEMA_CATALOG.exemplar_label_key_position WHERE metric_name=metric_name_text ) THEN
+        FOREACH k in ARRAY label_keys LOOP
+            INSERT INTO SCHEMA_CATALOG.exemplar_label_key_position VALUES (metric_name_text, k, current_position);
+            current_position := current_position + 1;
+        END LOOP;
+        RETURN QUERY (
+            SELECT row.metric_name, json_object_agg(row.key, row.position) FROM (
+                SELECT metric_name, key, pos as position FROM SCHEMA_CATALOG.exemplar_label_key_position
+                    WHERE metric_name=metric_name_text GROUP BY metric_name, key, pos ORDER BY pos
+            ) AS row GROUP BY row.metric_name LIMIT 1
+        );
+    RETURN;
+    END IF;
+
+    -- Position already exists for some keys. Let's add for the new keys only.
+    FOREACH k in ARRAY label_keys LOOP
+        IF (SELECT count(key) = 0 FROM SCHEMA_CATALOG.exemplar_label_key_position WHERE metric_name=metric_name_text AND key=k) THEN
+            SELECT max(pos) + 1 INTO new_position FROM SCHEMA_CATALOG.exemplar_label_key_position WHERE metric_name=metric_name_text;
+            RAISE NOTICE 'inserting key % at pos %', k, new_position;
+            INSERT INTO SCHEMA_CATALOG.exemplar_label_key_position VALUES (metric_name_text, k, new_position);
+            new_position := -1;
+        END IF;
+    END LOOP;
+    RETURN QUERY (
+    SELECT row.metric_name, json_object_agg(row.key, row.position) FROM (
+        SELECT metric_name, key, pos as position FROM SCHEMA_CATALOG.exemplar_label_key_position
+            WHERE metric_name=metric_name_text GROUP BY metric_name, key, pos ORDER BY pos
+        ) AS row GROUP BY row.metric_name
+    );
+END;
+$$
+LANGUAGE PLPGSQL;
+GRANT EXECUTE ON FUNCTION SCHEMA_CATALOG.get_or_create_exemplar_label_key_positions(TEXT, TEXT[]) TO prom_writer;
+GRANT EXECUTE ON FUNCTION SCHEMA_CATALOG.get_or_create_exemplar_label_key_positions(TEXT, TEXT[]) TO prom_reader;
+
+CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.get_or_create_exemplar_label_key_positions(metric_name_text TEXT)
```

Comment:
Should this be just get instead of get_or_create ? I don't see how this function can create positions.

Reply:
Yes, it should be only get.